### PR TITLE
Double precision: change default behaviour and many docs

### PR DESCRIPTION
**Pull request by PhilipVinc**
In the old netket we were using `float` and `complex` as dtypes everywhere.
However those are, in jax, known as weak dtypes (which means they are converted to single precision np.float32 as soon as they meet a single precision number).

As the samplers now by default give single-precision samples, this meant that computations were being done in single precision.

I changed to use np.float64 everywhere. Also the default of RbM and other models is now float64 instead of float32 to better match legacy behaviour.

Also add tons of warnings in the docs about this.

### CODE REVIEW

File: Examples/DissipativeIsing1d/ising1d.py [hunk 1]
```diff
@@ -34,7 +34,7 @@
 
 # Observables
 obs_sx = nk.operator.LocalOperator(hi)
-obs_sy = nk.operator.LocalOperator(hi, dtype=complex)
+obs_sy = nk.operator.LocalOperator(hi, dtype=np.complex128)
```
Comment (gcarleo) [inline Examples/DissipativeIsing1d/ising1d.py:37]: why do we need to pass numpy types now?

Comment (gcarleo) [reply Examples/DissipativeIsing1d/ising1d.py:37]: From the numpy docs 

> Note that, above, we use the Python float object as a dtype. NumPy knows that int refers to np.int_, bool means np.bool_, that float is np.float_ and complex is np.complex_.

Comment (gcarleo) [reply Examples/DissipativeIsing1d/ising1d.py:37]: I would therefore keep standard python data types, ```complex```, ```float``` etc also because it's not entirely transparent why ```LocalOperator``` would need a numpy data type

Comment (PhilipVinc) [reply Examples/DissipativeIsing1d/ising1d.py:37]: Yeah, this was just me doing find and replace. Here we don't need it.

In the other place (the model), it's written here: https://netket.github.io/netket/preview/double-prec/docs/sharp-bits.html#nans-in-training-and-loss-of-precision . It's due to jax's handling of `float` vs `np.float64` documented [here](https://jax.readthedocs.io/en/latest/type_promotion.html?highlight=type-promotion).
As I write in the docs, `float` is weak so gets converted to `float32` as soon as he meets a float32, instead of keeping the precision.

Comment (PhilipVinc) [reply Examples/DissipativeIsing1d/ising1d.py:37]: We don't really need those dtypes in the examples, as in this PR I am moving everything to strong double precision by default, but I am trying to give examples to users and push them to use `np.float**` by defaualt

Comment (gcarleo) [reply Examples/DissipativeIsing1d/ising1d.py:37]: I understand but this seems to me an issue with Flax, essentially, since numpy has already moved towards standardizing its data type to be consistent with the standard python ones...

Comment (gcarleo) [reply Examples/DissipativeIsing1d/ising1d.py:37]: for example in this table I can't see ```np_.float``` https://jax.readthedocs.io/en/latest/type_promotion.html?highlight=type-promotion so maybe it's also jax lagging a bit behind here?

**Comment by gcarleo** [comment]
In the old version if I remember correctly the precision for Jax was set globally to double precision once for all at include time or something... why can't we do that and simplify all of this?

File: Examples/DissipativeIsing1d/ising1d.py (hunk 1 above)
Comment (PhilipVinc) [reply Examples/DissipativeIsing1d/ising1d.py:37]: it's jax, not flax. it's not an issue, it's intended behaviour. 
>The promotion rules used by JAX for floating-point types are similar to those used by PyTorch.

>These differences are motivated by the fact that accelerator devices, such as GPUs and TPUs, either pay a significant performance penalty to use 64-bit floating point types (GPUs) or do not support 64-bit floating point types at all (TPUs). Classic NumPy’s promotion rules are too willing to overpromote to 64-bit types, which is problematic for a system designed to run on accelerators.

It's true that it's a bit surprising, which is why I am writing it in three places of the documentation.

We were not seeing those things before because we were handling parameters in a very unintended way, while now we are more consistent with jax.

**Comment by PhilipVinc** [comment]
I'm doing it.

**Comment by gcarleo** [comment]
> I'm doing it.

so you mean that despite telling jax to use globally double precision you still have this mixed precision issues?

**Comment by gcarleo** [comment]
my point is that single precision scientific computations are essentially meaningless :) so we should just default to double precision... but maybe I am being too drastic

**Comment by PhilipVinc** [comment]
The differences are the following:
before: everything was hardcoded to `np.float64` or `np.complex128`. Sampler output, operators, and the parameters of machines.

Now: the dtype of everything can be chosen. 
Operators default to `np.float64` unless necessary, basically behaving like numpy matrices.  
Samplers give `np.float32` samples unless the dtype is changed. 
Default models are `np.float64`.

If you want, those issues are arising for two reasons:
1) Before we were forcing the dtype of jax parameters to float64/complex128, now we are not coercing it anymore. This is an improvement, in my opinion (and I need it for density matrices, as one layer might be real and the other complex). This also allows us to take code written for jax/flax and use it with no changes.
2) Samplers now give samples in `float32` by default. This is senseful, to me, because we are simply storing numbers like `-1,0,1,2,3,4...` which don't need precision. So, now, because of (1), if you get a float32 input and pass it through a model with `float` dtype, while the parameters themselves are float64, the computation and the gradient will be float32. 

There are two solutions:
 1 - move samplers to default to `float64`. Let people use `float` with no problems. Smart people will realise they just need `float32` and it will be faster. but then they will need to learn they have to use `float64` instead of `float` in their code to make it work well together.
 2 - keep samplers `float32`, and teach users from the beginning to use the dtype they want, and why.

**Comment by PhilipVinc** [comment]
> so you mean that despite telling jax to use globally double precision you still have these mixed precision issues?

The flag does not do that. The flag simply enables double precision for jax. That's all (it's disabled by default). In Netket 2 this was synonymous to globally switching netket to double precision or not because we were coercing types to whatever jax thought `float` was.

**Comment by PhilipVinc** [comment]
>my point is that single precision scientific computations are essentially meaningless :) so we should just default to double precision... but maybe I am being too drastic

I think this is a drastic view. You need don't need double precision everywhere.

**Comment by gcarleo** [comment]
OK so... it seems that the main issue is the type used in samplers, if I understand correctly, since lower precision can give speed ups? But let me understand something: in samplers, do we use double precisions do compute acceptance rates etc even when we declare the samples to be single precision (or, why not, ```int```, that would be even better) ?

**Comment by gcarleo** [comment]
In other words, I agree with you that it makes totally sense to use lower precision for the sample types (again, ```int```in most cases is more than enough!) However, if a model is specified in double precision, I would expect that the sampler defaults to double precision to compute the acceptance, and that the gradient updates are computed in the same precision designed for the parameters of the models. If this is the current behavior I see no problem

**Comment by PhilipVinc** [comment]
Result of the discussion: move samplers to double precision.

**Comment by PhilipVinc** [comment]
@gcarleo tests pass, so it can be reviewed now

**Review by gcarleo — APPROVED** [review]
